juju/main.py

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+
+
+JUJU_CHARM_DIR = os.environ.get("JUJU_CHARM_DIR")
+if JUJU_CHARM_DIR is None:
+    # Set the module-level variable assuming
+    # $JUJU_CHARM_DIR/lib/juju/main.py structure.
+    JUJU_CHARM_DIR = os.path.dirname(os.path.dirname(
+        os.path.dirname(os.path.abspath(__file__))))
+
+if __name__ == '__main__':
+    # Use modules from the $JUJU_CHARM_DIR/lib directory (conditionally, to
+    # avoid side-effects on module import).
+    # An absolute path is used as using a symlink like
+    # hooks/install -> ../lib/juju/main.py
+    # will result in $JUJU_CHARM_DIR/lib/juju being the first path stored
+    # in sys.path during hook execution which is undesirable.
+    # This happens because the directory containing a script is added to
+    # sys.path by the interpreter which gets resolved if it is a symlink.
+    del sys.path[0]
+    sys.path.insert(0, os.path.join(JUJU_CHARM_DIR, 'lib'))
+
+from charm import Charm # noqa
```

Comment:
I think this is a symptom of not having the separation of concerns and direction of dependency right.  The framework code is expected to be reused across all charms, so it should be independent from the charm and have no dependencies on the charm.  Rather, the framework should be called by the charm and told what implementation to use.  Something like:

```python
import juju.framework

class MyCharm(juju.framework.CharmBase):
    ...

if __name__ == '__main__':
    juju.framework.main(MyCharm)
```

The charm is almost certainly going to want additional third-party libraries at some point, so we'll have to deal with the packaging issue sooner or later.  Once we do so, then it makes sense to me for the framework to be packaged as a library as well (also, components).

With this approach, the path issues are resolved automatically, since Python always knows how to find and import libraries, and the charm code has the right paths because it's the thing that's being executed.

That does leave wiring up at least the initial hook executable to symlink to the charm code, but I think we'll likely end up wanting some manner of build phase regardless, if only to pre-cache the libraries the charm depends on (currently done via the wheelhouse, but there are issues with that approach which need to be addressed) to minimize network traffic during install and handle the restricted network deployment case.  If we presume a build phase, then it seems reasonable to have it do a bit of work to wire up the Python-friendly charm code to the Juju-friendly hook code.

Reply:
I ended up introducing a couple of things to preserve the current behavior.

If it's a hook execution by a unit agent or a manual hook execution simulation (e.g. in `debug-hooks`), the following will run:

```
if __name__ == '__main__':
    setup_path()
    main()
```

Then in the `main` function I use `importlib` to load the necessary modules which are then used to set hook symlinks up on install/upgrade-charm and instantiate the Charm class.

https://github.com/canonical/charmbase/pull/6/files#diff-460b509ca15cacedde5eeb7bcba28b81R165-R167

As a result, `juju/main.py` can be imported without side-effects and will work as expected and set up `sys.path` in the `__name__ == '__main__'` case. If somebody uses `juju.main` as a module and call `juju.main.main` function, the expectation is that they will have the environment set up correctly.